File: Problem1.py
# Problem 1: Path Sum-II

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.all_paths = []

    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        self.helper(root, [], 0, targetSum)
        return self.all_paths

    def helper(self, root, currList, currSum, targetSum):
        # base
        if not root:
            return
        # logic

        # action
        currList.append(root.val)
        currSum += root.val

        if not root.left and not root.right:
            # this is a leaf node
            # if sum equals target add deep copy of current list to result
            # note: List is called by reference so remember to pop(backtrack)
            if currSum == targetSum:
                an_answer_list = []
                an_answer_list.extend(currList)
                self.all_paths.append(an_answer_list)
            # Optimizing leaf node checks: [Optional]
            currList.pop()
            return

        # recurse
        self.helper(root.left, currList, currSum, targetSum)
        self.helper(root.right, currList, currSum, targetSum)

        # Backtrack
        currList.pop()

"""
    Complexity Analysis:
    Time: O(n) [n - #nodes in the tree]
    Space: O(n) [size of stack = Height of tree]
    Did the code run successfully on Leetcode?: Yes
    Did you face any issues while soving this problem? : No, the lecture video explained the strategy clearly.
"""

File: Problem1.java
import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Problem1 {
    List<List<Integer>> result;
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        result = new ArrayList<>();
        helper(root, 0, new ArrayList<Integer>(), targetSum);
        return result;
    }
//     private void helper(TreeNode root, int currSum, List<Integer> path, int target){
//         // base
//         if(root == null) return;

//         // logic
//         currSum += root.val;
//         List<Integer> currPath = new ArrayList<>(path);
//         currPath.add(root.val);
//         if(root.left == null && root.right == null){
//             if(currSum == target){
//                 result.add(currPath);
//             }
//         }
//         helper(root.left, currSum, currPath, target);
//         helper(root.right, currSum, currPath, target);
//     }
    private void helper(TreeNode root, int currSum, List<Integer> path, int target){
        // base
        if(root == null) return;

        // logic
        currSum += root.val;
        path.add(root.val);
        if(root.left == null && root.right == null){
            if(currSum == target){
                result.add(new ArrayList<>(path));
            }
        }
        helper(root.left, currSum, path, target);

        helper(root.right, currSum, path, target);

        path.remove(path.size()-1);
    }
}